File: src/renotize.nim
import std/os
import std/json
import std/osproc
import std/strutils
import std/sequtils
import std/browsers
import std/strformat

import zippy/ziparchives
when isMainModule: import cligen

const rcodesignBin = staticRead("../rcodesign")

let rcodesignPath = getTempDir() / "rcodesign"
if execCmdEx(&"{rcodesignPath} -V").exitCode != 0:
  writeFile(rcodesignPath, rcodesignBin)
  setFilePermissions(rcodesignPath, {fpUserRead, fpUserWrite, fpUserExec})

type KeyboardInterrupt = object of CatchableError

proc handler() {.noconv.} =
  raise newException(KeyboardInterrupt, "Keyboard Interrupt")

setControlCHook(handler)

proc provision*() =
  ## Utility method to provision required information for notarization using a step-by-step process.
  # generate private key for code signing request
  discard execCmd("openssl genrsa -out private-key.pem 2048")

  # generate CSR
  discard execCmd(&"{rcodesignPath} generate-certificate-signing-request --pem-source private-key.pem --csr-pem-path csr.pem")

  # upload CSR to apple
  echo "This next step should be completed in the browser."
  echo "Press 'Enter' to open the browser and continue."
  discard readLine(stdin)
  openDefaultBrowser("https://developer.apple.com/account/resources/certificates/add")

  # print step by step instructions
  echo "1. Select 'Developer ID Application' as the certificate type"
  echo "2. Click 'Continue'"
  echo "3. Select the G2 Sub-CA (Xcode 11.4.1 or later) Profile Type"
  echo "4. Select 'csr.pem' using the file picker"
  echo "5. Click 'Continue'"
  echo "6. Click the 'Download' button to download your certificate"
  echo "8. Save the certificate next to the private-key.pem and csr.pem files"

  echo "Press 'Enter' when you have saved the certificate"
  discard readLine(stdin)

  let certFiles = walkFiles(getCurrentDir() / "*.cer").toSeq()
  if certFiles.len == 0:
    echo "No .cer file found in current directory"
    quit(1)

  echo "This next step should be completed in the browser."
  echo "Press 'Enter' to open the browser and continue."
  discard readLine(stdin)
  openDefaultBrowser("https://appstoreconnect.apple.com/access/users")

  echo "1. Click on 'Keys'"
  echo "2. If this is your first time, click on 'Request Access' and wait until it is granted"
  echo "3. Click on 'Generate API Key'"
  echo "4. Enter a name for the key"
  echo "5. For Access, select 'Developer'"
  echo "6. Click on 'Generate'"
  echo "7. Copy the Issuer ID and enter it here: ('Enter' to confirm)"
  let issuerId = readLine(stdin).strip()
  if issuerId.len == 0:
    echo "Issuer ID cannot be empty"
    quit(1)
  echo "8. Copy the Key ID and enter it here: ('Enter' to confirm)"
  let keyId = readLine(stdin).strip()
  if keyId.len == 0:
    echo "Key ID cannot be empty"
    quit(1)
  echo "7. Next to the entry of the newly-created key in the list, click on 'Download API Key'"
  echo "8. In the following pop-up, Click on 'Download'"
  echo "10. Save the downloaded .p8 file next to the private-key.pem and csr.pem files"

  echo "Press 'Enter' when you have saved the certificate"
  discard readLine(stdin)

  # find the first file ending in .p8
  let p8Files = walkFiles(getCurrentDir() / "*.p8").toSeq()
  if p8Files.len == 0:
    echo "No .p8 file found in current directory"
    quit(1)
  discard execCmdEx(&"{rcodesignPath} encode-app-store-connect-api-key -o app-store-key.json {issuerId} {keyId} {p8Files[0]}")

  echo "Success!"
  echo "You can now sign your app using these two files:"
  echo "  - private-key.pem"
  echo &"  - {certFiles[0]}"
  echo "You can also use this file to notarize your app:"
  echo "  - app-store-key.json"

proc unpackApp*(inputFile: string, outputDir = "") =
  ## Unpacks the given ZIP file to the target directory.
  var targetDir = outputDir
  if targetDir != "" and dirExists(targetDir):
    removeDir(targetDir)
  if targetDir == "":
    targetDir = inputFile
    removeSuffix(targetDir, ".zip")

  extractAll(inputFile, targetDir)

  let extractedFile = walkDirs(joinPath(targetDir, "*.app")).toSeq()[0]
  let newTargetDir = joinPath(splitPath(targetDir)[0], splitPath(extractedFile)[1])
  moveFile(extractedFile, newTargetDir)
  removeDir(targetDir)

proc signApp*(inputFile: string, keyFile: string, certFile: string) =
  ## Signs a .app bundle with the given Developer Identity.
  let entitlements = """<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd"><plist version="1.0"><dict><key>com.apple.security.cs.allow-unsigned-executable-memory</key><true/></dict></plist>"""

  writeFile("entitlements.plist", entitlements)

  discard execCmd(&"{rcodesignPath} sign -e entitlements.plist --pem-source {keyFile} --der-source {certFile} {inputFile}")

  removeFile("entitlements.plist")

proc notarizeApp*(inputFile: string, appStoreKeyFile: string): string =
  ## Notarizes a .app bundle with the given Developer Account and bundle ID.
  discard execCmd(&"{rcodesignPath} notary-submit --api-key-path {appStoreKeyFile} --staple {inputFile}")

proc packDmg*(
  inputFile: string,
  outputFile: string,
  volumeName = "",
) =
  ## Packages a .app bundle into a .dmg file.
  var vName = volumeName
  if volumeName == "":
    vName = splitFile(inputFile).name
  let cmd = &"hdiutil create -fs HFS+ -format UDBZ -ov -volname {vName} -srcfolder {inputFile} {outputFile}"
  discard execShellCmd(cmd)

proc signDmg*(inputFile: string, keyFile: string, certFile: string) =
  ## Signs a .dmg file with the given Developer Identity.
  discard execCmd(&"{rcodesignPath} sign -e entitlements.plist --pem-source {keyFile} --der-source {certFile} {inputFile}")

proc notarizeDmg*(inputFile: string, appStoreKeyFile: string): string =
  ## Notarizes a .dmg file with the given Developer Account and bundle ID.
  discard execCmd(&"{rcodesignPath} notary-submit --api-key-path {appStoreKeyFile} --staple {inputFile}")

proc status*(uuid: string, appStoreKeyFile: string): string =
  ## Checks the status of a notarization operation given its UUID.
  let data = parseJson(execProcess(&"{rcodesignPath} notary-log --api-key-path {appStoreKeyFile} {uuid}"))

  var status = "not started"
  if "notarization-info" in data:
    status = data["notarization-info"]["Status"].getStr()

  return status

proc fullRun*(inputFile, keyFile, certFile, appStoreKeyFile: string) =
  # Programmatic interface for the full run operation to allow
  # dynamically passing in configuration data from memory at runtime.
  echo "Unpacking app"
  unpackApp(inputFile)

  let appFile = walkDirs(joinPath(splitPath(inputFile)[0], "*.app")).toSeq()[0]

  echo "Signing app"
  signApp(appFile, keyFile, certFile)

  echo "Notarizing app"
  echo notarizeApp(appFile, appStoreKeyFile)

  echo "Signing stapled app"
  signApp(appFile, keyFile, certFile)

  let (dir, name, _) = splitFile(appFile)
  let dmgFile = &"{joinPath(dir, name)}.dmg"

  echo "Packing DMG"
  packDmg(appFile, dmgFile)

  echo "Signing DMG"
  signDmg(dmgFile, keyFile, certFile)

  echo "Notarizing DMG"
  echo notarizeDmg(dmgFile, appStoreKeyFile)

  echo "Cleaning up"
  removeFile("*-app.zip")
  removeDir("extracted distro")

  echo "Done"

proc fullRunCli*(inputFile: string, keyFile = "", certFile = "", appStoreKeyFile = "") =
  ## Fully notarize a given .app bundle, creating a signed
  ## and notarized artifact for distribution.
  var
    keyFileInt: string
    certFileInt: string
    appStoreKeyFileInt: string

  if keyFile == "":
    keyFileInt = getEnv("RN_KEY_FILE")
  else:
    keyFileInt = keyFile
  if certFile == "":
    certFileInt = getEnv("RN_CERT_FILE")
  else:
    certFileInt = certFile
  if appStoreKeyFile == "":
    appStoreKeyFileInt = getEnv("RN_APP_STORE_KEY_FILE")
  else:
    appStoreKeyFileInt = appStoreKeyFile

  if keyFileInt == "" or certFileInt == "" or appStoreKeyFileInt == "":
    echo "No configuration data was found via config file or environment."
    quit(1)

  fullRun(inputFile, keyFileInt, certFileInt, appStoreKeyFileInt)

when isMainModule:
  dispatchMulti(
    [provision],
    [unpackApp, cmdName="unpack-app", help = {
        "input_file": "The path to the ZIP file containing the .app bundle.",
        "output_dir": "The directory to extract the .app bundle to.",
    }],
    [signApp, cmdName="sign-app", help = {
        "input_file": "The path to the .app bundle.",
        "keyFile": "The private key generated via the 'provision' command.",
        "certFile": "The certificate file obtained via the 'provision' command.",
    }],
    [notarizeApp, cmdName="notarize-app", help = {
        "input_file": "The path to the .app bundle.",
        "appStoreKeyFile": "The app-store-key.json file obtained via the 'provision' command.",
    }],
    [packDmg, cmdName="pack-dmg", help = {
        "input_file": "The path to the .app bundle.",
        "output_file": "The name of the DMG file to write to.",
        "volume_name": "The name to use for the DMG volume. By default the base name of the input file."
    }],
    [signDmg, cmdName="sign-dmg", help = {
        "input_file": "The path to the .dmg file.",
        "keyFile": "The private key generated via the 'provision' command.",
        "certFile": "The certificate file obtained via the 'provision' command.",
    }],
    [notarizeDmg, cmdName="notarize-dmg", help = {
        "input_file": "The path to the .dmg file.",
        "appStoreKeyFile": "The app-store-key.json file obtained via the 'provision' command.",
    }],
    [status, help = {
        "uuid": "The UUID of the notarization operation.",
        "appStoreKeyFile": "The app-store-key.json file obtained via the 'provision' command.",
    }],
    [fullRunCli, cmdName = "full-run", help = {
        "input_file": "The path to the the ZIP file containing the .app bundle.",
        "keyFile": "The private key generated via the 'provision' command.",
        "certFile": "The certificate file obtained via the 'provision' command.",
        "appStoreKeyFile": "The app-store-key.json file obtained via the 'provision' command.",
    }],
  )

File: src/renotize.rs
use crate::common::zip_dir;
use anyhow::{Result, anyhow};
use app_store_connect::notary_api::{self, SubmissionResponseStatus};
use apple_codesign::{
    AppleCodesignError, CodeSignatureFlags, NotarizationUpload, Notarizer, SettingsScope,
    SigningSettings, UnifiedSigner,
    cli::{
        certificate_source::{CertificateDerSigningKey, CertificateSource, PemSigningKey},
        config::SignConfig,
    },
    stapling::Stapler,
};
use indicatif::{ProgressBar, ProgressStyle};
use itertools::Itertools;
use jwalk::WalkDir;
use plist::Value;
use rsa::{RsaPrivateKey, pkcs1::EncodeRsaPrivateKey};
use std::{
    fs::{self, File},
    io::Cursor,
    path::{Path, PathBuf},
    process::Command,
    time::Duration,
};
use x509_certificate::X509CertificateBuilder;
use zip::CompressionMethod;

const APPLE_TIMESTAMP_URL: &str = "http://timestamp.apple.com/ts01";
const ENTITLEMENTS_PLIST: &str = r#"<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd"><plist version="1.0"><dict><key>com.apple.security.cs.allow-unsigned-executable-memory</key><true/></dict></plist>"#;

fn notarize_file(
    input_file: &Path,
    app_store_key_file: &Path,
    staple_file: Option<&Path>,
) -> Result<()> {
    let notarizer = Notarizer::from_api_key(app_store_key_file)?;

    println!("Uploading file to notarization service");
    let upload = notarizer.notarize_path(input_file, None)?;

    match upload {
        NotarizationUpload::UploadId(id) => {
            println!("Upload complete");
            println!("Notarization UUID: {id}");
            println!("Waiting for notarization to complete");

            let wait_limit = Duration::from_secs(1800);
            let wait_interval = Duration::from_secs(5);
            let start_time = std::time::Instant::now();

            let bar =ProgressBar::new(wait_limit.as_secs()).with_style(
                ProgressStyle::with_template(
                    "{bar:48.green/black} {human_pos:>5.green}/{human_len:<5.green} {per_sec:.red} eta {eta:.blue}",
                )
                .unwrap()
                .progress_chars("━╾╴─"),
            );

            bar.inc(1);

            let mut status;

            loop {
                status = notarizer.get_submission(&id)?;

                let elapsed = start_time.elapsed();

                if status.data.attributes.status != notary_api::SubmissionResponseStatus::InProgress
                {
                    println!("Notary API Server has finished processing the uploaded asset");
                    break;
                }

                if elapsed >= wait_limit {
                    println!("reached wait limit after {}s", elapsed.as_secs());
                    break;
                }

                std::thread::sleep(wait_interval);
                bar.inc(wait_interval.as_secs());
            }

            bar.finish();

            if status.data.attributes.status == SubmissionResponseStatus::Accepted {
                println!("Notarization accepted");

                println!("Stapling notarization to file");
                let stapler = Stapler::new()?;

                match staple_file {
                    Some(sf) => stapler.staple_path(sf)?,
                    None => stapler.staple_path(input_file)?,
                }
            } else {
                println!("Notarization failed.");

                let log = notarizer.fetch_notarization_log(&id)?;

                for line in serde_json::to_string_pretty(&log)?.lines() {
                    println!("notary log> {line}");
                }
            }
        }
        NotarizationUpload::NotaryResponse(_) => {
            unreachable!("NotarizationUpload::NotaryResponse should be returned")
        }
    }

    Ok(())
}

pub fn unpack_app(input_file: &Path, output_dir: &Path, bundle_id: &str) -> Result<PathBuf> {
    if output_dir.exists() {
        std::fs::remove_dir_all(output_dir)?;
    }
    std::fs::create_dir_all(output_dir)?;

    let app_zip = fs::read(input_file)?;
    zip_extract::extract(Cursor::new(app_zip), output_dir, false)?;

    let mut app_path = None;
    for entry in fs::read_dir(output_dir)? {
        let entry = entry?;
        let path = entry.path();
        if let Some(ext) = path.extension()
            && ext.to_string_lossy() == "app"
        {
            let info_plist_path = path.join("Contents/Info.plist");
            app_path = Some(path);
            let mut info_plist = Value::from_file(&info_plist_path)?
                .into_dictionary()
                .ok_or(anyhow!("Info.plist is not a dictionary"))?;
            info_plist.insert(
                "CFBundleIdentifier".to_string(),
                plist::Value::String(bundle_id.to_owned()),
            );
            Value::Dictionary(info_plist).to_file_xml(&info_plist_path)?;
        }
    }

    Ok(app_path.unwrap())
}

pub fn sign_app(input_file: &Path, key_file: &Path, cert_file: &Path) -> Result<()> {
    let pem_key = PemSigningKey {
        paths: vec![key_file.to_path_buf()],
    };

    let der_key = CertificateDerSigningKey {
        paths: vec![cert_file.to_path_buf()],
    };

    let mut sign_config = SignConfig::default();
    sign_config.signer.pem_path_key = Some(pem_key);
    sign_config.signer.certificate_der_key = Some(der_key);

    let mut settings = SigningSettings::default();

    let certs = sign_config.signer.resolve_certificates(false)?;
    certs.load_into_signing_settings(&mut settings)?;

    if let Some(team_id) = settings.set_team_id_from_signing_certificate() {
        println!("Inferred team ID: {team_id}");
    }

    settings.set_time_stamp_url(APPLE_TIMESTAMP_URL)?;

    settings.set_entitlements_xml(SettingsScope::Main, ENTITLEMENTS_PLIST)?;

    settings.set_code_signature_flags(SettingsScope::Main, CodeSignatureFlags::RUNTIME);

    for path in [
        "Contents/MacOS/zsync",
        "Contents/MacOS/zsyncmake",
        "Contents/MacOS/python",
        "Contents/MacOS/pythonw",
    ] {
        settings.set_code_signature_flags(
            SettingsScope::Path(path.to_string()),
            CodeSignatureFlags::RUNTIME,
        );
    }

    let signer = UnifiedSigner::new(settings);

    println!("Signing bundle at {}", input_file.display());
    signer.sign_path_in_place(input_file)?;

    Ok(())
}

pub fn notarize_app(input_file: &Path, app_store_key_file: &Path) -> Result<()> {
    notarize_file(input_file, app_store_key_file, None)
}

pub fn pack_dmg(input_file: &Path, output_file: &Path, volume_name: &Option<String>) -> Result<()> {
    let volume_name = if let Some(name) = volume_name {
        name.clone()
    } else {
        let input_file_name = input_file
            .file_name()
            .ok_or(anyhow!("Input file name is not valid UTF-8"))?
            .to_string_lossy();
        input_file_name
            .strip_suffix(".app")
            .ok_or(anyhow!("Input file name does not end with .app"))?
            .to_string()
    };

    println!("Name: {volume_name}");

    let mut cmd = Command::new("hdiutil");
    cmd.args([
        "create",
        "-fs",
        "HFS+",
        "-format",
        "UDBZ",
        "-ov",
        "-volname",
        &volume_name,
        "-srcfolder",
        &input_file.to_string_lossy(),
        &output_file.to_string_lossy(),
    ]);
    let status = cmd.status()?;

    if !status.success() {
        anyhow::bail!("Unable to create DMG.");
    }

    Ok(())
}

pub fn sign_dmg(input_file: &Path, key_file: &Path, cert_file: &Path) -> Result<()> {
    let pem_key = PemSigningKey {
        paths: vec![key_file.to_path_buf()],
    };

    let der_key = CertificateDerSigningKey {
        paths: vec![cert_file.to_path_buf()],
    };

    let mut sign_config = SignConfig::default();
    sign_config.signer.pem_path_key = Some(pem_key);
    sign_config.signer.certificate_der_key = Some(der_key);

    let mut settings = SigningSettings::default();

    let certs = sign_config.signer.resolve_certificates(false)?;
    certs.load_into_signing_settings(&mut settings)?;

    if let Some(team_id) = settings.set_team_id_from_signing_certificate() {
        println!("Automatically setting team ID from signing certificate: {team_id}");
    }

    let signer = UnifiedSigner::new(settings);

    signer.sign_path_in_place(input_file)?;

    Ok(())
}

pub fn notarize_dmg(input_file: &Path, app_store_key_file: &Path) -> Result<()> {
    notarize_file(input_file, app_store_key_file, None)
}

pub fn pack_zip(input_file: &Path, output_file: &Path) -> Result<()> {
    let mut files = WalkDir::new(input_file).into_iter();

    let file = File::create(output_file).unwrap();

    zip_dir(
        &mut files,
        Some(&input_file.parent().unwrap().to_owned()),
        file,
        CompressionMethod::Deflated,
    )?;

    Ok(())
}

pub fn notarize_zip(input_file: &Path, app_store_key_file: &Path, app_file: &Path) -> Result<()> {
    notarize_file(input_file, app_store_key_file, Some(app_file))
}

pub fn status(uuid: &str, app_store_key_file: &Path) -> Result<()> {
    let notarizer = Notarizer::from_api_key(app_store_key_file)?;

    let Ok(log) = notarizer.fetch_notarization_log(uuid) else {
        println!("Status not available yet.");
        return Ok(());
    };

    let status = match log.get("status") {
        Some(status) => status.as_str().unwrap(),
        None => "unknown",
    };
    println!("Status: {status}");

    if let Some(issues) = log.get("issues")
        && let serde_json::Value::Array(_) = issues
    {
        let issues = issues.as_array().unwrap().iter().map(|issue| {
            let issue = issue.as_object().unwrap();
            let message = issue.get("message").unwrap().as_str().unwrap();
            let doc_url = issue.get("docUrl").unwrap().as_str().unwrap();
            let path = issue.get("path").unwrap().as_str().unwrap();
            (message, doc_url, path)
        });

        for (key, group) in &issues.chunk_by(|(message, _, _)| *message) {
            println!("Error: {key}");
            for (i, (_, doc_url, path)) in group.enumerate() {
                if i == 0 {
                    println!("Documentation: {doc_url}\nAffected files:");
                }
                println!("  - {path}");
            }
        }
    }

    Ok(())
}

pub fn full_run(
    input_file: &Path,
    bundle_id: &str,
    key_file: &Path,
    cert_file: &Path,
    app_store_key_file: &Path,
    create_zip: bool,
    create_dmg: bool,
) -> Result<()> {
    let is_zip = input_file.extension().unwrap() == "zip";

    let app_path = if is_zip {
        let output_dir = input_file.with_extension("");
        println!("Unpacking app to {}", output_dir.display());
        unpack_app(input_file, &output_dir, bundle_id)?
    } else {
        input_file.to_path_buf()
    };

    println!("Signing app at {}", app_path.display());
    sign_app(&app_path, key_file, cert_file)?;

    println!("Notarizing app at {}", app_path.display());
    notarize_app(&app_path, app_store_key_file)?;

    if create_zip {
        let zip_path = app_path
            .parent()
            .unwrap()
            .with_file_name(format!(
                "{}-notarized",
                input_file.file_stem().unwrap().to_string_lossy()
            ))
            .with_extension("zip");
        println!("Packing ZIP to {}", input_file.display());
        pack_zip(&app_path, &zip_path)?;

        if is_zip {
            fs::remove_file(input_file)?;
        }
        fs::rename(&zip_path, input_file.with_extension("zip"))?;

        println!("Notarizing ZIP at {}", input_file.display());
        notarize_zip(input_file, app_store_key_file, &app_path)?;
    }

    if std::env::consts::OS == "macos" {
        if create_dmg {
            let dmg_path = input_file.with_extension("dmg");
            println!("Packing DMG to {}", dmg_path.display());
            pack_dmg(&app_path, &dmg_path, &None)?;
            println!("Signing DMG at {}", dmg_path.display());
            sign_dmg(&dmg_path, key_file, cert_file)?;
            println!("Notarizing DMG at {}", dmg_path.display());
            notarize_dmg(&dmg_path, app_store_key_file)?;

            fs::remove_dir_all(app_path.parent().unwrap())?;
        }
    } else {
        println!("Skipping DMG creation and signing: Only supported on macOS.");
    }

    println!("Done!");

    Ok(())
}

pub fn provision() -> Result<()> {
    let cert_dir = Path::new("certificates");
    fs::create_dir_all(cert_dir)?;

    let mut rng = rand::thread_rng();
    let priv_key = RsaPrivateKey::new(&mut rng, 2048)?;

    let data = priv_key.to_pkcs1_pem(rsa::pkcs8::LineEnding::LF)?;
    let private_key_path = cert_dir.join("private-key.pem");
    fs::write(&private_key_path, data)?;

    let pem_key = PemSigningKey {
        paths: vec![private_key_path],
    };

    let cert_source = CertificateSource {
        pem_path_key: Some(pem_key),
        ..Default::default()
    };

    let certs = cert_source.resolve_certificates(false)?;
    let private_key = certs.private_key()?;

    let mut builder = X509CertificateBuilder::default();
    builder
        .subject()
        .append_common_name_utf8_string("Apple Code Signing CSR")
        .map_err(|e| AppleCodesignError::CertificateBuildError(format!("{e:?}")))?;

    println!("Generating CSR. You may be prompted to enter credentials to unlock the signing key.");
    let pem = builder
        .create_certificate_signing_request(private_key.as_key_info_signer())?
        .encode_pem()?;

    let csr_path = cert_dir.join("csr.pem");

    std::fs::write(csr_path, pem.as_bytes())?;

    println!("This step should be completed in the browser.");
    println!("Press 'Enter' to open the browser and continue.");

    let mut input = String::new();
    std::io::stdin().read_line(&mut input)?;

    webbrowser::open("https://developer.apple.com/account/resources/certificates/add")?;

    println!("1. Select 'Developer ID Application' as the certificate type");
    println!("2. Click 'Continue'");
    println!("3. Select the G2 Sub-CA (Xcode 11.4.1 or later) Profile Type");
    println!("4. Select 'csr.pem' using the file picker");
    println!("5. Click 'Continue'");
    println!("6. Click the 'Download' button to download your certificate");
    println!("7. Save the certificate next to the private-key.pem and csr.pem files:");
    println!(
        "     {}/developerID_application.cer",
        cert_dir.to_string_lossy()
    );

    println!("Press 'Enter' when you have saved the certificate");

    let mut input = String::new();
    std::io::stdin().read_line(&mut input)?;

    let cert_file = cert_dir.join("developerID_application.cer");
    loop {
        if cert_file.exists() {
            break;
        }
        println!("Certificate not found. Press 'Enter' when you have saved the certificate.");
        println!(
            "Make sure to name the file 'developerID_application.cer' and save it next to the private-key.pem and csr.pem files."
        );
        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
    }

    println!("Success!");

    println!("This step should be completed in the browser.");
    println!("Press 'Enter' to open the browser and continue.");

    let mut input = String::new();
    std::io::stdin().read_line(&mut input)?;

    webbrowser::open("https://appstoreconnect.apple.com/access/users")?;

    println!("1. Click on 'Keys'");
    println!(
        "2. If this is your first time, click on 'Request Access' and wait until it is granted"
    );
    println!("3. Click on 'Generate API Key'");
    println!("4. Enter a name for the key");
    println!("5. For Access, select 'Developer'");
    println!("6. Click on 'Generate'");
    println!("7. Copy the Issuer ID and enter it here: ('Enter' to confirm)");

    let mut issuer_id = String::new();

    loop {
        std::io::stdin().read_line(&mut issuer_id)?;

        if issuer_id.len() > 1 {
            break;
        }

        println!("Issuer ID can not be empty, please try again.");
    }

    println!("8. Copy the Key ID and enter it here: ('Enter' to confirm)");

    let mut key_id = String::new();

    loop {
        std::io::stdin().read_line(&mut key_id)?;

        if key_id.len() > 1 {
            break;
        }

        println!("Key ID can not be empty, please try again.");
    }

    println!(
        "9. Next to the entry of the newly-created key in the list, click on 'Download API Key'"
    );
    println!("10. In the following pop-up, Click on 'Download'");
    println!("11. Save the downloaded .p8 file next to the private-key.pem and csr.pem files.");

    let app_store_cert_path;

    loop {
        if let Some(path) = fs::read_dir(cert_dir)?.find_map(|entry| {
            let entry = entry.unwrap();
            let path = entry.path();
            match path.extension() {
                Some(ext) => {
                    if ext == "p8" {
                        Some(path)
                    } else {
                        None
                    }
                }
                None => None,
            }
        }) {
            app_store_cert_path = path;
            break;
        }
        println!("Key file not found. Press 'Enter' when you have saved the key file.");
        println!("Make sure to save the file next to the private-key.pem and csr.pem files.");
        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
    }

    let unified = app_store_connect::UnifiedApiKey::from_ecdsa_pem_path(
        &issuer_id,
        &key_id,
        &app_store_cert_path,
    )?;

    let app_store_key_path = cert_dir.join("app-store-key.json");
    unified.write_json_file(app_store_key_path)?;

    println!("Success!");
    println!("You can now sign your app using these three files:");
    println!("  - private-key.pem");
    println!("  - app-store-key.pem");
    println!("  - {}", cert_file.file_name().unwrap().to_string_lossy());

    Ok(())
}
